### PR TITLE
Shadow models and predictions for membership attack

### LINKED ISSUE #1756: Support overlapping shadow model datasets
**Is your feature request related to a problem? Please describe.**
The current implementation of shadow model training creates disjoint train and test datasets for each shadow model: 
shadow_dataset_size = len(x) // len(self._shadow_models)
This does not enable using high numbers of shadow models, especially with small datasets.

**Describe the solution you'd like**
The original Shokri et al. paper says in their experimental setup section: "There is no overlap between the datasets of the target model and those of the shadow models, but the datasets used for different shadow models can overlap with each other". 
I suggest to add an optional parameter to this class indicating whether to allow overlapping datasets (with default True).

### PR DESCRIPTION
**Pull request by abigailgold**
# Description
1. Enable creating shadow datasets that overlap (as opposed to only supporting disjoint datasets). At the moment only supports completely disjoint or complete overlap.
2. Enable sending the blackbox membership attack model predictions as input (instead of the model itself). This enables supporting use cases where the model or training data (X) is not available. 

Fixes #1756 

## Type of change
- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing
Added test for new functionality (both for shadow models and membership attack)

**Test Configuration**:
- OS: 12.4
- Python version: 3.8
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/regression/pytorch.py [hunk 1]
```diff
@@ -533,6 +533,45 @@ def compute_loss(  # type: ignore # pylint: disable=W0221
 
         return loss.detach().cpu().numpy()
 
+    def compute_loss_from_predictions(  # type: ignore # pylint: disable=W0221
```
Comment (beat-buesser) [inline art/estimators/regression/pytorch.py:536]: Why do we need this new method in addition to the existing method `compute_losses`?

File: art/estimators/estimator.py [hunk 2]
```diff
@@ -338,6 +338,16 @@ def compute_loss(self, x: np.ndarray, y: Any, **kwargs) -> np.ndarray:
         """
         raise NotImplementedError
 
+    def compute_loss_from_predictions(self, pred: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
```
Comment (beat-buesser) [inline art/estimators/estimator.py:341]: Why do we need this new method in addition to the existing method compute_losses?

File: art/estimators/regression/pytorch.py (hunk 1 above)
Comment (beat-buesser) [inline art/estimators/regression/pytorch.py:536]: Why do we need this new method in addition to the existing method `compute_losses`?

File: art/estimators/estimator.py (hunk 2 above)
Comment (abigailgold) [reply art/estimators/estimator.py:341]: To serve the use case where we may not have access to training data (features) but only to the model predictions.

**Review by beat-buesser — APPROVED** [review]
Hi @abigailgold Thank you very much for your pull request! It looks good to me.

**Pull request merged**